Short-Answer/Algorithms_Answers.md

```diff
@@ -1 +1,22 @@
-Add your answers to the Algorithms exercises here.
\ No newline at end of file
+Add your answers to the Algorithms exercises here.
+
+Exercise I
+
+a) O(n) - I think this is linear since it starts at n^2 to get to n^3. The running time will increase with the size of the input.
+
+b) O(n^4) - because there are 4 nested loops, each with O(n).
```

Comment:
```
b)  sum = 0                                   O(1)
    for i in range(n):                        O(n)
      i += 1                                  O(1)
      for j in range(i + 1, n):               O(n), still
        j += 1                                O(1)
        for k in range(j + 1, n):             O(n), still
          k += 1                              O(1)
          for l in range(k + 1, 10 + k):      O(10)
            l += 1                            O(1)
            sum += 1                          O(1)
```

The `j` and `k` loops don't go all the way to `n`, but they go some fraction of the way to `n`. And fractional constants are dropped.

The `l` loop is tricky. Close examination shows that it is:

```
for (l = 1 + k; l < 10 + k; l++)
```

but `k` doesn't change throughout a single loop of `l`. So it's effectively constant and the loop is equivalent to:

```
for (l = 1; l < 10; l++)     O(10) == O(1)
```

No matter what `k` you plug into that loop, it always cycles 10 times.

Since the loops are all nested, we multiply the values.

```
O(1) +  <-- this is for sum = 0
O(n) * O(n) * O (n) * O(1) <-- for nested loops

Simplifying:

O(1) + O(n*n*n*1)
O(1) + O(n^3)
O(n^3)               O(n^3) dominates
```

Answer: `O(n^3)`